providers/enchant_nuspell.cpp

```diff
@@ -0,0 +1,444 @@
+ * License along with this library; if not, write to the
+ * Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
+ * Boston, MA 02110-1301, USA.
+ *
+ * In addition, as a special exception, Dom Lachowicz
+ * gives permission to link the code of this program with
+ * non-LGPL Spelling Provider libraries (eg: a MSFT Office
+ * spell checker backend) and distribute linked combinations including
+ * the two.  You must obey the GNU General Public License in all
+ * respects for all of the code used other than said providers.  If you modify
+ * this file, you may extend this exception to your version of the
+ * file, but you are not obligated to do so.  If you do not wish to
+ * do so, delete this exception statement from your version.
+ */
+
+/*
+ * This is the Nuspell Enchant Backend.
+ * Nuspell is by Dimitrij Mijoski and Sander van Geloven.
+ * See: http://nuspell.github.io/
+ */
+
+#include "config.h"
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+
+#include <string>
+#include <vector>
+
+#include "enchant-provider.h"
+#include "unused-parameter.h"
+
+#include <nuspell/dictionary.hxx>
+#include <nuspell/finder.hxx>
+
+#include <glib.h>
+
+using namespace std;
+using namespace nuspell;
+
+/***************************************************************************/
+
+class NuspellChecker
+{
+public:
+	bool checkWord (const char *word, size_t len);
+	char **suggestWord (const char* const word, size_t len, size_t *out_n_suggs);
+
+	bool requestDictionary (const char * szLang);
+
+private:
+	Dictionary nuspell;
+};
+
+/***************************************************************************/
+
+bool
+NuspellChecker::checkWord(const char *utf8Word, size_t len)
+{
+	// the 8bit encodings use precomposed forms
+	char *normalizedWord = g_utf8_normalize (utf8Word, len, G_NORMALIZE_NFC);
+
+	if (nuspell.spell(string(normalizedWord)))
+		return true;
+	else
+		return false;
+}
+
+char**
+NuspellChecker::suggestWord(const char* const utf8Word, size_t len, size_t *nsug)
+{
+	// the 8bit encodings use precomposed forms
+	char *normalizedWord = g_utf8_normalize (utf8Word, len, G_NORMALIZE_NFC);
+	auto suggestions = vector<string>();
+	nuspell.suggest(string(normalizedWord), suggestions);
+	if (suggestions.empty())
+		return nullptr;
+	*nsug = suggestions.size();
+	char **sug = g_new0 (char *, *nsug + 1);
+	size_t i = 0;
+	for (auto& suggest : suggestions) {
+		char *word = g_new0(char, suggest.size() + 1);
+		strcpy(word, suggest.c_str());
+		sug[i] = word;
+		i++;
+	}
+	return sug;
+}
+
+static void
+s_buildDictionaryDirs (vector<string> & dirs)
+{
+	dirs.clear ();
+
+	/* 1. personal overrides for Enchant
+	 *    ~/.config/enchant/nuspell
+	 */
+	gchar * tmp;
+	char * config_dir = enchant_get_user_config_dir ();
+	tmp = g_build_filename (config_dir, "nuspell", nullptr);
+	dirs.push_back (tmp);
+	free (config_dir);
+	g_free(tmp);
+
+	/* Dynamically retieved from Nuspell dictionary finder:
+	 * 2. personal overides for Hunspell
```

Comment:
typo: should be "overrides"

Reply:
Fixed in https://github.com/nuspell/enchant/commit/d4634184fb8ddcab5020b5ec6d5f162e4fca23f3#diff-b4a4b7f548e6522299f21f4f5632ff1cR121